src/components/Form.tsx

```diff
@@ -0,0 +1,265 @@
+				if (handlers.onBlur) handlers.onBlur(args[1]);
+				break;
+			case 'onInitialized':
+				if (handlers.onInitialized) handlers.onInitialized();
+				break;
+			default:
+				break;
+		}
+	}
+};
+
+const createWebformInstance = async (
+	FormConstructor: FormConstructor | undefined,
+	formSource: FormSource,
+	element: MutableRefObject<HTMLDivElement>['current'],
+	props: FormProps,
+) => {
+	const {
+		formioform,
+		Formio,
+		src,
+		url,
+		form,
+		formReady,
+		onFormReady,
+		options = {},
+		submission,
+		...handlers
+	} = props;
+	if (!options?.events) {
+		options.events = getDefaultEmitter();
+	}
+	const promise = FormConstructor
+		? new FormConstructor(element, formSource, options)
+		: new FormClass(element, formSource, options);
+	const instance = await promise.ready;
+	if (Object.keys(handlers).length > 0) {
+		instance.onAny((...args: [string, ...any[]]) =>
+			onAnyEvent(handlers, ...args),
+		);
+	}
+	return instance;
+};
+
+// Define effective props (aka I want to rename these props but also maintain backwards compatibility)
+const getEffectiveProps = (props: FormProps) => {
+	const {
+		Formio: FormioProp,
+		formioform,
+		form,
+		src,
+		formReady,
+		onFormReady,
+	} = props;
+	const formConstructor = FormioProp !== undefined ? FormioProp : formioform;
+
+	const formSource = form !== undefined ? form : src;
+
+	const formReadyCallback =
+		onFormReady !== undefined ? onFormReady : formReady;
+
+	return { formConstructor, formSource, formReadyCallback };
+};
+
+const Form = (props: FormProps) => {
+	const [formInstance, setFormInstance] = useState<Webform | null>(null);
+	const renderElement = useRef<HTMLDivElement | null>(null);
+	const prevFormDefinition = useRef<JSON | null>(null);
+
+	const { formConstructor, formSource, formReadyCallback } =
+		getEffectiveProps(props);
+	const { submission, url } = props;
+
+	useEffect(() => () => formInstance?.destroy(), [formInstance]);
+
+	useEffect(() => {
+		if (renderElement.current === null) {
+			console.warn('Form element not found');
+			return;
+		}
+		if (typeof formSource === 'string') {
+			createWebformInstance(
+				formConstructor,
+				formSource,
+				renderElement.current,
+				props,
+			).then((instance) => {
+				if (instance) {
+					instance.src = formSource;
+					if (formReadyCallback) formReadyCallback(instance);
+					setFormInstance(instance);
+				}
+			});
+		} else if (
+			typeof formSource === 'object' &&
+			!isEqual(formSource, prevFormDefinition.current)
+		) {
+			prevFormDefinition.current = cloneDeep(formSource);
+			createWebformInstance(
+				formConstructor,
+				formSource,
+				renderElement.current,
+				props,
+			).then((instance) => {
+				if (instance && formSource) {
+					instance.form = formSource;
+					if (url) {
+						instance.url = url;
+					}
+					if (formReadyCallback) formReadyCallback(instance);
+					setFormInstance(instance);
+				}
+			});
+		}
+	}, [formConstructor, formSource, props, url, formReadyCallback]);
```

Comment:
@brendanbond  
I presume we have added props to take care of any props change.

I believe it also include handlers aka onAnyEvent functions(onSubmit etc..), which does not require recreating form instance.
Recreating the form instance for every props changes(specifically handlers) can lead to unnecessary performance overhead.
We only need to register them in form instance and turn off previous events on handlers change.

Ideal would be to add specific prop and create another effect for handler.

Let me know if I need to add snippet and testing for it.

Reply:
Yeah that's a good point - I'll make an update